File: cameo-api-cpp/src/cameo/impl/RequesterImpl.cpp
/*
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under the EUPL, Version 1.1 only (the "License");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * http://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */

#include "RequesterImpl.h"
#include "../Application.h"
#include "../Serializer.h"
#include "ServicesImpl.h"
#include "RequestSocketImpl.h"
#include "../message/JSON.h"
#include "../message/Message.h"
#include <sstream>

using namespace std;

namespace cameo {

const std::string RequesterImpl::REQUESTER_PREFIX = "req.";
std::mutex RequesterImpl::m_mutex;
int RequesterImpl::m_requesterCounter = 0;

RequesterImpl::RequesterImpl(application::This * application, const std::string& url, int requesterPort, int responderPort, const std::string& name, int responderId, int requesterId) :
	m_application(application),
	m_requesterPort(requesterPort),
	m_name(name),
	m_responderId(responderId),
	m_requesterId(requesterId),
	m_canceled(false) {

	stringstream repEndpoint;
	repEndpoint << url << ":" << responderPort;
	m_responderEndpoint = repEndpoint.str();

	// Create the request socket.
	m_requestSocket = m_application->createRequestSocket(m_responderEndpoint);

	// Create a socket REP.
	m_repSocket.reset(new zmq::socket_t(m_application->m_impl->m_context, ZMQ_REP));
	stringstream reqEndpoint;
	reqEndpoint << "tcp://*:" << m_requesterPort;

	m_repSocket->bind(reqEndpoint.str().c_str());
}

RequesterImpl::~RequesterImpl() {
	terminate();
}

int RequesterImpl::newRequesterId() {

	lock_guard<mutex> lock(m_mutex);
	m_requesterCounter++;

	return m_requesterCounter;
}

std::string RequesterImpl::getRequesterPortName(const std::string& name, int responderId, int requesterId) {

	stringstream requesterPortName;
	requesterPortName << REQUESTER_PREFIX << name << "." << responderId << "." << requesterId;

	return requesterPortName.str();
}

WaitingImpl * RequesterImpl::waiting() {
	return new GenericWaitingImpl(bind(&RequesterImpl::cancel, this));
}

void RequesterImpl::sendBinary(const std::string& requestData) {

	json::StringObject request;
	request.pushKey(message::TYPE);
	request.pushInt(message::REQUEST);

	request.pushKey(message::Request::APPLICATION_NAME);
	request.pushString(m_application->getName());

	request.pushKey(message::Request::APPLICATION_ID);
	request.pushInt(m_application->getId());

	request.pushKey(message::Request::SERVER_URL);
	request.pushString(m_application->getUrl());

	request.pushKey(message::Request::SERVER_PORT);
	request.pushInt(m_application->getPort());

	request.pushKey(message::Request::REQUESTER_PORT);
	request.pushInt(m_requesterPort);

	m_requestSocket->request(request.toString(), requestData);
}

void RequesterImpl::send(const std::string& requestData) {

	// encode the data
	string result;
	serialize(requestData, result);
	sendBinary(result);
}

void RequesterImpl::sendTwoBinaryParts(const std::string& requestData1, const std::string& requestData2) {

	json::StringObject request;
	request.pushKey(message::TYPE);
	request.pushInt(message::REQUEST);

	request.pushKey(message::Request::APPLICATION_NAME);
	request.pushString(m_application->getName());

	request.pushKey(message::Request::APPLICATION_ID);
	request.pushInt(m_application->getId());

	request.pushKey(message::Request::SERVER_URL);
	request.pushString(m_application->getUrl());

	request.pushKey(message::Request::SERVER_PORT);
	request.pushInt(m_application->getPort());

	request.pushKey(message::Request::REQUESTER_PORT);
	request.pushInt(m_requesterPort);

	m_requestSocket->request(request.toString(), requestData1, requestData2);
}

bool RequesterImpl::receiveBinary(std::string& response) {

	unique_ptr<zmq::message_t> message(new zmq::message_t);
	m_repSocket->recv(message.get(), 0);

	// Get the JSON request.
	json::Object request;
	json::parse(request, message.get());

	int type = request[message::TYPE].GetInt();

	if (type == message::RESPONSE) {
		// Get the second part for the message.
		message.reset(new zmq::message_t);
		m_repSocket->recv(message.get(), 0);
		response = string(message->data<char>(), message->size());
	}
	else if (type == message::CANCEL) {
		m_canceled = true;
	}

	// Create the reply.
	string data = "OK";
	size_t size = data.length();
	unique_ptr<zmq::message_t> reply(new zmq::message_t(size));
	memcpy(reply->data(), data.c_str(), size);

	m_repSocket->send(*reply);

	return !m_canceled;
}

bool RequesterImpl::receive(std::string& data) {

	string bytes;
	bool result = receiveBinary(bytes);

	parse(bytes, data);

	return result;
}

void RequesterImpl::cancel() {

	stringstream requesterEndpoint;
	requesterEndpoint << m_application->getUrl() << ":" << m_requesterPort;

	json::StringObject request;
	request.pushKey(message::TYPE);
	request.pushInt(message::CANCEL);

	// Create a request socket only for the request.
	unique_ptr<RequestSocketImpl> requestSocket = m_application->createRequestSocket(requesterEndpoint.str());
	requestSocket->request(request.toString());
}

void RequesterImpl::terminate() {

	if (m_repSocket.get() != nullptr) {
		m_repSocket.reset(nullptr);

		bool success = m_application->removePort(getRequesterPortName(m_name, m_responderId, m_requesterId));
		if (!success) {
			cerr << "Server cannot destroy requester " << m_name << endl;
		}
	}

	m_requestSocket.reset();
}

}

File: cameo-api/src/main/java/fr/ill/ics/cameo/impl/RequesterImpl.java
/*
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under the EUPL, Version 1.1 only (the "License");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * http://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */

package fr.ill.ics.cameo.impl;

import java.util.concurrent.atomic.AtomicInteger;

import org.json.simple.JSONObject;
import org.json.simple.parser.ParseException;

import fr.ill.ics.cameo.Zmq;
import fr.ill.ics.cameo.messages.JSON;
import fr.ill.ics.cameo.messages.Message;
import fr.ill.ics.cameo.strings.Endpoint;

public class RequesterImpl {

	public static final String REQUESTER_PREFIX = "req.";

	private ThisImpl application;
	Zmq.Context context;
	private int requesterPort;
	private String name;
	private int responderId;
	
	// Need for a unique id per Application instance.
	private int requesterId;
	private static AtomicInteger requesterCounter = new AtomicInteger();
	
	private Zmq.Socket requester;
	private RequestSocket requestSocket;
	
	private boolean canceled = false;
	private RequesterWaitingImpl waiting = new RequesterWaitingImpl(this);
		
	public RequesterImpl(ThisImpl application, Zmq.Context context, String url, int requesterPort, int responderPort, String name, int responderId, int requesterId) {
		this.application = application;
		this.context = context;
		this.requesterPort = requesterPort;
		String responderEndpoint = url + ":" + responderPort;
		this.name = name;
		this.responderId = responderId;
		this.requesterId = requesterId;

		// Create the REQ socket.
		requestSocket = application.createRequestSocket(responderEndpoint);
		
		// Create the REP socket.
		requester = context.createSocket(Zmq.REP);
		requester.bind("tcp://*:" + requesterPort);
		
		waiting.add();
	}
	
	public String getName() {
		return name;
	}
	
	public static int newRequesterId() {
		return requesterCounter.incrementAndGet();
	}
	
	public static String getRequesterPortName(String name, int responderId, int requesterId) {
		return REQUESTER_PREFIX + name + "." + responderId + "." + requesterId;
	}
		
	public void send(byte[] requestData) {
		
		JSONObject request = new JSONObject();
		request.put(Message.TYPE, Message.REQUEST);
		request.put(Message.Request.APPLICATION_NAME, application.getName());
		request.put(Message.Request.APPLICATION_ID, application.getId());
		request.put(Message.Request.SERVER_URL, application.getEndpoint().getProtocol() + "://" + application.getEndpoint().getAddress());
		request.put(Message.Request.SERVER_PORT, application.getEndpoint().getPort());
		request.put(Message.Request.REQUESTER_PORT, requesterPort);
		
		Zmq.Msg message = application.message(request);
		
		// Set request in the next frame.
		message.add(requestData);
		
		requestSocket.request(message);
	}
	
	public void send(String request) {
		send(Message.serialize(request));
	}
	
	public void sendTwoParts(byte[] requestData1, byte[] requestData2) {
		
		JSONObject request = new JSONObject();
		request.put(Message.TYPE, Message.REQUEST);
		request.put(Message.Request.APPLICATION_NAME, application.getName());
		request.put(Message.Request.APPLICATION_ID, application.getId());
		request.put(Message.Request.SERVER_URL, application.getEndpoint().getProtocol() + "://" + application.getEndpoint().getAddress());
		request.put(Message.Request.SERVER_PORT, application.getEndpoint().getPort());
		request.put(Message.Request.REQUESTER_PORT, requesterPort);
		
		Zmq.Msg message = application.message(request);
		
		// Set request1 and request2 in the next frames.
		message.add(requestData1);
		message.add(requestData2);
		
		requestSocket.request(message);
	}

	public byte[] receive() {
		
		Zmq.Msg message = null;
		
		try {
			message = Zmq.Msg.recvMsg(requester);

			if (message == null) {
				return null;
			}
			
			// Get the JSON request object.
			JSONObject request = application.parse(message);
			
			// Get the type.
			long type = JSON.getLong(request, Message.TYPE);
						
			if (type == Message.RESPONSE) {
				return message.getLastData();
			}
			else if (type == Message.CANCEL) {
				canceled = true;
				return null;
			}
			else {
				return null;
			}
		}
		catch (ParseException e) {
			System.err.println("Cannot parse message");
			return null;
		}
		finally {
			if (message != null) {
				message.destroy();
			}
			
			// Send to the responder
			Zmq.Msg reply = new Zmq.Msg();
			reply.add("OK");
			reply.send(requester);
		}
	}

	public String receiveString() {
		return Message.parseString(receive());
	}
	
	public void cancel() {
		
		Endpoint endpoint = application.getEndpoint().withPort(requesterPort);

		JSONObject request = new JSONObject();
		request.put(Message.TYPE, Message.CANCEL);
		
		// Create the request socket. We can create it here because it should be called only once.
		RequestSocket requestSocket = application.createRequestSocket(endpoint.toString());
		requestSocket.request(application.message(request));
		
		// Terminate the socket.
		requestSocket.terminate();
	}

	public boolean isCanceled() {
		return canceled;
	}
	
	public void terminate() {
		
		waiting.remove();
		
		// Terminate the request socket.
		requestSocket.terminate();
		
		context.destroySocket(requester);
		
		try {
			application.removePort(getRequesterPortName(name, responderId, requesterId));
			
		} catch (Exception e) {
			System.err.println("Cannot terminate requester: " + e.getMessage());
		}
	}
	
	@Override
	public String toString() {
		return REQUESTER_PREFIX + name + "." + requesterId + ":" + application.getName() + "." + application.getId() + "@" + application.getEndpoint();
	}
	
}